python/lsst/analysis/drp/makeQaTractTables.py

```diff
@@ -0,0 +1,418 @@
+        if not catList:
+            raise RuntimeError("No catalogs read for {}.".format(inputObjCats))
+        allCats = pd.concat(catList, axis=0)
+        # The object "id" is associated with the dataframe index.  Add a
+        # column that is the id so that it is available for operations on it,
+        # e.g. cat["id"].
+        allCats["id"] = allCats.index
+        # Reindex to match sorting of objTable_tract tables
+        inputObjectTable = inputObjectTable.get(parameters={"columns": []})
+        allCats = allCats.reindex(inputObjectTable.index)
+        return pipeBase.Struct(qaTractTable=allCats)
+
+
+class MakeForcedQaTractTablesTaskConnections(BaseMakeQaTractTablesTaskConnections,
+                                             defaultTemplates={"coaddName": "deep"},
+                                             dimensions=("tract", "skymap")):
+    pass
+
+
+class MakeForcedQaTractTablesTaskConfig(BaseMakeQaTractTablesTaskConfig,
+                                        pipelineConnections=MakeForcedQaTractTablesTaskConnections):
+    # We want the following to come from the *_meas catalogs as they reflect
+    # what happened in SFP calibration.
+    columnsToCopyFromMeas = pexConfig.ListField(
+        dtype=str,
+        default=["calib_", ],
+        doc=("List of string \"prefixes\" to identify the columns to copy.  All columns with names "
+             "that start with one of these strings will be copied from the *_meas catalogs into the "
+             "*_forced_src catalogs UNLESS the full column name contains one of the strings listed "
+             "in the notInColumnStrList config.")
+    )
+    # We want the following to come from the *_ref catalogs as they reflect
+    # the forced measurement states.
+    columnsToCopyFromRef = pexConfig.ListField(
+        dtype=str,
+        default=["detect_", "merge_peak_sky", "merge_measurement_", ],
+        doc=("List of string \"prefixes\" to identify the columns to copy.  All columns with names "
+             "that start with one of these strings will be copied from the *_ref catalogs into the "
+             "*_forced_src catalogs UNLESS the full column name contains one of the strings listed "
+             "in the notInColumnStrList config.")
+    )
+
+
+class MakeForcedQaTractTablesTask(BaseMakeQaTractTablesTask):
+    """Read in {coaddName}Coadd_obj tables and aggregate QA columns by tract.
+
+    The values of interest here are those from the forced catalogs (i.e.
+    "forced_src") and, in this case, certain columns from the "meas" and "ref"
+    catalogs that don't get propagated to the "forced_src" catalogs are added
+    to the persisted table, which has dataset type "qaTractTable_forced".
+
+    See BaseMakeQaTractTablesTask for full description.
+    """
+
+    ConfigClass = MakeForcedQaTractTablesTaskConfig
+    _DefaultName = "makeForcedQaTractTablesTask"
+
+    colsToLoadList = None
+    refColsToLoadList = None
+    measColsToLoadList = None
```

Comment:
Your instincts that said we might not want to use class attributes like this were unfortunately correct, and I don't see a super easy fix; logic that basically says "do something special the first time through a loop" is pretty much always hard clean up.

This "I need to read one catalog in order to get the schema" problem is actually exactly why we have a practice of writing separate `_schema` datasets for all of our afw table catalogs, and I think this is a pretty clear sign that we need to be doing the same for our parquet tables.  That would let us load the schemas in `__init__`, compute the columns there, and make these variables frozen instance attributes.  That actually lets us chain schema-definition through multiple tasks - we can compute the final schema for some downstream task before actually running data through the first ones.

We won't get that done on this ticket, though (I'll open a new one for it).  Instead, let's use a pair of small helper class; something like this:

```
class ForcedQaLoadList:
     def __init__(self, inputCatCols, activeFilterName):
        self.forcedCols = ...  # compute cols to load from forced
        self.refCols = ... # compute cols to load from ref
        self.measCols = ... # compute cols to load from meas

class UnforcedQaLoadList:
    def __init__(self, inputCatCols, activeFilterName):
        self.cols = ...  # compute cols to load from meas
```
and a new abstract method:
```
@abstractmethod
def makeLoadList(self, inputCatCols, activeFilterName):
    """Make a "load list" instance appropriate for this derived class.
        ...
    """
        raise NotImplementedError()
```
The derived classes would implement that to just construct one of those helper classes.

The base class `run` method would then do something like this:
```
loadList = None
for inputCat in inputObjCats:
    ...
    if loadList is None:
        loadList = self.makeLoadList(inputCatCols, activeFilterName)
    cat = self.collateTable(..., loadList)
```
i.e. the base class would be responsible for figuring out when to make the load lists, but not what they contain, and it would then pass it into `collateTable` (which would need a new argument for it, but then _might_ be able to subtract some existing ones).

Reply:
First, a huge thanks for the master class you are giving me here...truly appreciated (and will hopefully pay in dividends for future reviewers of my code...if I can manage to retain any of the lessons learned here!)  So, I think I've got all your suggestions above implemented.  My current bad-practice speculation is my passing in the class `config`s to the new helper classes (seemed easier than accessing the individual config parameters and passing them in...but does it still make sense to have these `config`s defined where they are?)

Reply:
Passing in the `config` objects isn't that bad, and I think the config definitions themselves are fine now.  But I do think it'd be _slightly_ better to pass the config parameters separately instead; there are competing goals:
 - make code concise
 - don't pass a function _more_ than what it (or some interface it's trying to match) needs (in this case, all of the other config options those classes shouldn't care about).

I'd give the second of those more weight, generally, but this is the sort of thing I wouldn't have even noticed if you hadn't brought it up first.